File: Reversing.Kr/PEPassword/get_key.cs
using System;
using System.Threading;
using System.Threading.Tasks;

class Program
{
  public static void Main (string[] args) {
    UInt32[] lastP = new UInt32[4]{0x2be16e72, 0x5a08c120, 0xad5b56ca, 0x07b3a312};
    for (int i = 0; i < 3; i++)
    {
      Console.WriteLine("{0:X8}", lastP[i]);
      Parallel.For(0x0, 0xffffffff + 1, (j) => {
        if (NextP(lastP[i], j) == lastP[i+1]) {
          Console.WriteLine("q = {0:X8} next_q = {1:X8}", j, next_q(lastP[i], j));
        }
      });
    }
  }

  public UInt32 NextP(UInt32 p, UInt32 q) {
    q = Rol(q, p & 0x000000ff);
    p = p ^ q;
    p = Ror(p, (q & 0x0000ff00) >> 8);
    return p;
  }

  public UInt32 NextQ(UInt32 p, UInt32 q) {
    q = Rol(q, p & 0x000000ff);
    p = p ^ q;
    p = Ror(p, (q & 0x0000ff00) >> 8);
    q = q + p;
    return q;
  }

  public UInt32 Rol(UInt32 data, UInt32 shift, UInt32 size=32) {
    shift %= size;
    UInt32 remains = data >> (size - shift);
    UInt32 body = (data << shift) - (remains << size);
    return body + remains;
  }

  public UInt32 Ror(UInt32 data, UInt32 shift, UInt32 size=32) {
    shift %= size;
    UInt32 body = data >> shift;
    UInt32 remains = (data << (size - shift)) - (body << size);
    return body + remains;
  }
}

File: Reversing.Kr/PEPassword/getkey_c/getkey.c
#include <stdio.h>
#define DWORD unsigned long

DWORD toLittleEndian(DWORD dword)
{
  return (((dword >>  0) & 0xff) << 24)
        | (((dword >>  8) & 0xff) << 16)
        | (((dword >> 16) & 0xff) <<  8)
        | (((dword >> 24) & 0xff) <<  0);
}

DWORD l(DWORD data)
{
  return data & 0x000000ff;
}

DWORD h(DWORD data)
{
  return (data & 0x0000ff00) >> 8;
}

DWORD rol(DWORD data, int shift)
{
  int size = 32;
  return (data << shift) | (data >> (size - shift));
}

DWORD ror(DWORD data, int shift)
{
  int size = 32;
  return (data >> shift) | (data << (size - shift));
}

DWORD NextP(DWORD P, DWORD Q)  
{
  Q = rol(Q, (int)l(P));
  P = P ^ Q;
  P = ror(P, (int)h(Q));
  Q = Q + P;
  return P;
}  
  
DWORD NextQ(DWORD P, DWORD Q)  
{  
  Q = rol(Q, (int)l(P));
  P = P ^ Q;
  P = ror(P, (int)h(Q));
  Q = Q + P;
  return Q;
}

int main()
{
  DWORD originalTextArr[0x1000];
  DWORD packedTextArr[0x1000];
  DWORD p[0x1000];
  FILE *originalFp = fopen("..\\Original.exe", "rb");
  FILE *packedFp = fopen("..\\Packed.untouched.exe", "rb");
  fseek(originalFp, 0x1000, SEEK_SET);
  fseek(packedFp, 0x1000, SEEK_SET);
  for(int i=0; i<0x1000; i++) {
    DWORD dwordFromOriginal = 0x0;
    DWORD dwordFromPacked = 0x0;
    fread(&dwordFromOriginal, sizeof(DWORD), 1, originalFp);
    fread(&dwordFromPacked, sizeof(DWORD), 1, packedFp);
    dwordFromOriginal = toLittleEndian(dwordFromOriginal);
    dwordFromPacked = toLittleEndian(dwordFromPacked);
    p[i] = dwordFromOriginal ^ dwordFromPacked;
    p[i] = toLittleEndian(p[i]);
  }
  fclose(originalFp);
  fclose(packedFp);
  for(int i=0x0; i<0xD; i++) {
    printf("P = 0x%08X NextP = 0x%08X\n", p[i], p[i+1]);
    for(__int64 j=0; j<0xffffffff; j++) {
      if(NextP(p[i], (DWORD)j) == p[i+1]) {
        printf("Q = 0x%08X NextQ = 0x%08X\n", (DWORD)j, NextQ(p[i], (DWORD)j));
      }
    }
  }
  return 0;
}
